runtime/lua/vim/diagnostic.lua

```diff
@@ -871,7 +871,7 @@ local function set_list(loclist, opts)
   end
 
   if open then
-    if qf_id then
+    if not loclist then
```

Comment:
I think this it the wrong fix? The next line of code uses `qf_id` to figure out the index of the diagnostics quickfix list, and I don't know what passing `id = nil` would do (probably select the current quickfix list, which isn't necessarily the right thing).

Instead, we should compute `qf_id` right after creating a new quickfix. That is, above right after `setqflist`, we should have something like:

```lua
    -- Get the id of the newly created quickfix list.
    if qf_id == nil then
      qf_id = vim.fn.getqflist({ id = 0 }).id
    end
```

That will ensure that `qf_id` is always set if we're dealing with the quickfix list.

If you wanted to add a test for this, I think you'd do something like:

1. Create a non-diagnostics quickfix list, open it.
2. Call `vim.diagnostic.setqflist({ open = true })`
3. Assert that the currently opened/selected quickfix list is the qf list from step 2.

I believe that with the code you have here, it would still have the qf list from step 1 open.

Reply:
Unless I'm missing something, I believe this is fine. Let me explain my reasoning here:

- if `qf_id` is `nil`, then that means we've not found an existing diagnostics quickfix
- for that reason, the call to `vim.fn.setqflist` would have created a new quickfix list and made it active
- `vim.fn.getqflist({ id = qf_id, nr = 0 }).nr` will return the `nr` of the current quickfix list, which is the one we just created

- if `qf_id` is not `nil`, then that means we've found an existing diagnostics quickfix
- the call to `vim.fn.setqflist` with `'u'` would have updated it, instead of creating a new one
  - note that this does not make the diagnostics quickfix active, which is why we need to call `api.nvim_command(nr .. 'chistory')`
- `vim.fn.getqflist({ id = qf_id, nr = 0 }).nr` will return the `nr` of the existing diagnostics quickfix

> If you wanted to add a test for this, I think you'd do something like:
> Create a non-diagnostics quickfix list, open it.
> Call vim.diagnostic.setqflist({ open = true })
> Assert that the currently opened/selected quickfix list is the qf list from step 2.

When we call `vim.fn.setqflist` with the new diagnostics quickfix, it's also set as active as a new one is created. That means that even when `open=false`, if a quickfix list is already open, the new diagnostics quickfix list will be displayed as it's also active.
When `open=true`, `vim.fn.getqflist({ id = qf_id, nr = 0 }).nr` will return the `nr` of the newly created diagnostics quickfix, as it's been activated when it was created.

I've added a new test to describe this behavior: `sets new diagnostics quickfix as active`. Together with the other tests, it seems to me it covers the behavior we want to assert.